### PR TITLE
refactor(gatsby): refactor pagination in preparation to querying lmdb directly

### PR DESCRIPTION
**Pull request by vladar**
## Description

This is a refactoring PR in preparation to querying LMDB directly (using secondary indexes). It affects pagination, nodeModel and fast filters methods.

### Context

Currently our pagination needs full result after `filter + sort` operation to do the work. Basically if the result contains `100,000` items and we only need `20`, pagination expects an array of all `100,000` items.

It is not a problem at all for in-memory store but when working with any database you don't want to fetch `100,000` items to discard `99,980` of them in pagination.

### Changes in this PR

 Essentially, we need to change the way our current pagination works to support sliced subset of data returned from datastore. All changes below are needed to make it happen:

1. Now `limit` and `offset` are applied in datastore itself (not pagination resolver)
2. `totalCount` must become a part of query result (simple `results.length` call in pagination resolver won't work)
3. Deprecated `nodeModel.runQuery` because it *always* returns full array (i.e. `100,000` items even if you need `20`) and sadly it is a public API
4. Introduced its successor methods: `findAll` and `findOne` that return already sliced results and also got rid of `firstOnly` creeping in everywhere
5. Applied changes to resolvers related to pagination

### CODE REVIEW

File: packages/gatsby/src/datastore/in-memory/run-fast-filters.ts [hunk 1]
```diff
@@ -315,51 +317,51 @@ function collectBucketForElemMatch(
...
-  return sortNodes(result, sort, resolvedFields, stats)
+  const sortedResult = sortNodes(result, sort, resolvedFields, stats)
+  const totalCount = sortedResult.length
+
+  const entries =
+    skip || limit
+      ? sortedResult.slice(skip, limit ? skip + (limit ?? 0) : undefined)
+      : sortedResult
+
+  return { entries, totalCount }
```
Comment (vladar) [inline packages/gatsby/src/datastore/in-memory/run-fast-filters.ts:352]: The key change: applying `skip`/`limit` in the datastore itself. And providing `totalCount` separately.

File: packages/gatsby/src/datastore/in-memory/run-fast-filters.ts [hunk 2]
```diff
@@ -406,9 +404,6 @@ function convertAndApplyFastFilters(
     if (stats) {
       stats.totalIndexHits++
     }
-    if (firstOnly) {
-      return result.slice(0, 1)
-    }
```
Comment (vladar) [inline packages/gatsby/src/datastore/in-memory/run-fast-filters.ts:411]: Now the calling side must pass `limit: 1` (and empty `sort`) instead of `firstOnly`.

File: packages/gatsby/src/datastore/types.ts [hunk 3]
```diff
@@ -9,6 +9,11 @@ export interface ILmdbDatabases {
   nodesByType: Database<NodeId, NodeType>
 }
 
+export interface IQueryResult {
+  entries: Iterable<IGatsbyNode>
+  totalCount: number | (() => Promise<number>)
+}
```
Comment (vladar) [inline packages/gatsby/src/datastore/types.ts:15]: `totalCount` will be expensive for `lmdb` case, so we support a `thunk` option too.

File: packages/gatsby/src/schema/node-model.js [hunk 4]
```diff
@@ -222,15 +222,29 @@ class LocalNodeModel {
...
+    // TODO: show deprecation warning in v4
+    // reporter.warn(
+    //   `nodeModel.runQuery() is deprecated. Use nodeModel.findAll() or nodeModel.findOne() instead`
+    // )
+    if (args.firstOnly) {
+      return this.findOne(args, pageDependencies)
+    }
+    const { skip, limit, ...query } = args.query
+    const result = await this.findAll({ ...args, query }, pageDependencies)
+    return Array.from(result.entries)
```
Comment (vladar) [inline packages/gatsby/src/schema/node-model.js:243]: This is the main reason to deprecate `runQuery` - right now it ignores `limit`/`skip` and always returns full array of results.

File: packages/gatsby/src/schema/node-model.js [hunk 5]
```diff
@@ -307,35 +320,75 @@ class LocalNodeModel {
...
+   * @returns {Promise<Node | null>}
+   */
+  async findOne(args, pageDependencies = {}) {
+    const { query } = args
+    if (query.sort?.fields?.length > 0) {
+      // If we support sorting and return the first node based on sorting
+      // we'll have to always track connection not an individual node
+      reporter.warn(
+        `nodeModel.findOne() does not support sorting. Use nodeModel.findAll({ query: { limit: 1 } }) instead`
+      )
```
Comment (vladar) [inline packages/gatsby/src/schema/node-model.js:369]: A subtle nuance. Right now you can do this in your custom resolver:

```js
const firstNode = await nodeModel.runQuery({
  query: { sort: { fields: [`foo`] } }
  firstOnly: true
})
```

But it will silently ignore sorting under the hood. So you won't get "the first node sorted by `foo`". After this change it will fail with this error. Not sure if we should consider it a breaking change or not - I can remove it for now and keep silently ignoring sort.

File: packages/gatsby/src/schema/resolvers.ts [hunk 6]
```diff
@@ -75,34 +54,56 @@ export function findOne<TSource, TArgs>(
...
     const group = getProjectedField(info, `group`)
     const distinct = getProjectedField(info, `distinct`)
     const max = getProjectedField(info, `max`)
+    const min = getProjectedField(info, `min`)
+    const sum = getProjectedField(info, `sum`)
+
+    // Apply paddings for pagination
+    // (for previous/next node and also to detect if there is a previous/next page)
+    const skip = typeof args.skip === `number` ? Math.max(0, args.skip - 1) : 0
+    const limit = typeof args.limit === `number` ? args.limit + 1 : undefined
```
Comment (vladar) [inline packages/gatsby/src/schema/resolvers.ts:77]: A common trick when using databases to figure out if there is "next"/"previous" page is to request one additional item on both sides. This way you can avoid expensive `count` calls.

File: packages/gatsby/src/schema/resolvers.ts [hunk 7]
```diff
@@ -75,34 +54,56 @@ export function findOne<TSource, TArgs>(
...
+        tracer: context.tracer,
+      },
+      { path: context.path, connectionType: typeName }
     )
-    return paginate(result, { skip: args.skip, limit: args.limit })
+    return paginate(result, {
+      resultOffset: skip,
+      skip: args.skip,
+      limit: args.limit,
+    })
```
Comment (vladar) [inline packages/gatsby/src/schema/resolvers.ts:106]: Now that results are already sliced during `findAll` call, pagination can't just do `result.slice(skip, offset)` because it has no idea where `skip` points now. So we pass additional information on the current result offset.

With this information, we can calculate the actual slice start point: `skip - resultOffset`

File: packages/gatsby/src/schema/node-model.js (hunk 5 above)
Comment (wardpeet) [reply packages/gatsby/src/schema/node-model.js:369]: why can't we run findAll for our users ourselves? As a developer I would find it weird to do a findAll(limit: 1) instead of using findOne

File: packages/gatsby/src/schema/node-model.js [hunk 8]
```diff
@@ -307,35 +320,75 @@ class LocalNodeModel {
...
+      // TODO: throw in v4
+      // throw new Error(
+      //   `nodeModel.findOne() does not support sorting. Use nodeModel.findAll({ query: { limit: 1 } }) instead`
+      // )
+    }
+    const { gqlType, entries } = await this._query({
+      ...args,
+      query: { ...query, skip: 0, limit: 1, sort: undefined },
+    })
+    const result = Array.from(entries)
```
Comment (wardpeet) [inline packages/gatsby/src/schema/node-model.js:379]: What dataType is entries? Is it an iterator? Does it already only have 1 item?

If it's an iterator you could also do `entries.next()` to get the next value.

File: packages/gatsby/src/schema/resolvers.ts [hunk 9]
```diff
@@ -75,34 +54,56 @@ export function findOne<TSource, TArgs>(
...
-
-    const result = await findMany<TSource, PaginatedArgs<TArgs>>(typeName)(
-      source,
-      extendedArgs,
-      context,
-      info
+    if (context.stats) {
+      context.stats.totalRunQuery++
+      context.stats.totalPluralRunQuery++
+    }
```
Comment (wardpeet) [inline packages/gatsby/src/schema/resolvers.ts:92]: Do you know why we have these stats? Maybe add a comment why it exists? (I know it's from the old code but I like to know :joy:)

File: packages/gatsby/src/schema/resolvers.ts [hunk 10]
```diff
@@ -244,38 +245,52 @@ export const group: GatsbyResolver<
...
-    results = []
+export function paginate(
+  results: IQueryResult,
+  params: { skip?: number; limit?: number; resultOffset?: number }
+): IGatsbyConnection<IGatsbyNode> {
+  const { resultOffset = 0, skip = 0, limit } = params
+  if (resultOffset > skip) {
+    throw new Error("Result offset cannot be greater than `skip` argument")
   }
+  let countOrThunk = results.totalCount
```
Comment (wardpeet) [inline packages/gatsby/src/schema/resolvers.ts:267]: When is this a thunk?

File: packages/gatsby/src/schema/resolvers.ts [hunk 11]
```diff
@@ -244,38 +245,52 @@ export const group: GatsbyResolver<
...
+  const allItems = Array.from(results.entries)
 
-  const count = results.length
-  const items = results.slice(skip, limit && skip + limit)
+  const start = skip - resultOffset
+  const items = allItems.slice(start, limit && start + limit)
```
Comment (wardpeet) [inline packages/gatsby/src/schema/resolvers.ts:271]: can't this grow pretty big?

File: packages/gatsby/src/schema/type-definitions.ts [hunk 12]
```diff
@@ -26,7 +26,7 @@ export type GatsbyResolver<TSource, TArgs = { [argName: string]: any }> = (
 ) => any
 
 export interface IGatsbyConnection<NodeType> {
-  totalCount: number
+  totalCount: number | (() => Promise<number>)
```
Comment (wardpeet) [inline packages/gatsby/src/schema/type-definitions.ts:29]: So when using LMDB this will always be async and with in-memory it will be sync, right? Shouldn't we break this in v4 so it's always async?

I imagine that on the React side of things, graphql will resolve the promise, right?

File: packages/gatsby/src/schema/node-model.js (hunk 5 above)
Comment (vladar) [reply packages/gatsby/src/schema/node-model.js:369]: `findOne` registers data dependency for a single node. `findAll` registers data dependency for the whole connection. So in the first case query will be invalidated only when this specific node changes but in the second case - when _any_ node of this type changes.

We want to have as many individual node dependencies but not connection dependencies. So ideally if you don't care about sorting - `findOne` is the best choice for warm / inc.builds.

File: packages/gatsby/src/schema/node-model.js (hunk 8 above)
Comment (vladar) [reply packages/gatsby/src/schema/node-model.js:379]: It is iterable with a single item (because of `limit: 1`). So technically we can do `entries[Symbol.iterator]().next().value` here but I don't see much benefits given there is only one item anyway.

File: packages/gatsby/src/schema/resolvers.ts (hunk 10 above)
Comment (vladar) [reply packages/gatsby/src/schema/resolvers.ts:267]: It will be a thunk in `lmdb` case

File: packages/gatsby/src/schema/resolvers.ts (hunk 11 above)
Comment (vladar) [reply packages/gatsby/src/schema/resolvers.ts:271]: Well, essentially it happens before this PR - you get 100,000 items in results from fast filters and then slice it here in pagination (e.g. get a slice of 20 items only). But this array contains references to nodes which are already in-memory. Given each reference is ~8 bytes, then slice with 1000 elements will take ~8kb additional memory.

With this PR slicing occurs in the datastore itself. In pagination you get up to 22 items (20 + 1 item before and 1 after). 
Also for lmdb case this `results` will be an iterable and slicing will be lazy as you traverse (in theory - because we do multiple traversals over those results anyways and it might be actualy cheaper to load all 20 items in memory at once).

File: packages/gatsby/src/schema/type-definitions.ts (hunk 12 above)
Comment (vladar) [reply packages/gatsby/src/schema/type-definitions.ts:29]: Yes, `graphql` resolves this promise. But also promises are rather slow so if we can return sync result from the resolver - we'd better do this (maybe not in this case but as a rule of thumb)

File: packages/gatsby/src/schema/resolvers.ts (hunk 9 above)
Comment (vladar) [reply packages/gatsby/src/schema/resolvers.ts:92]: We add it to telemetry here:

https://github.com/gatsbyjs/gatsby/blob/7c343ffd803c6ada0602678b9a6e6d164a58a4fb/packages/gatsby/src/commands/build.ts#L180-L184

File: packages/gatsby/src/schema/node-model.js (hunk 5 above)
Comment (wardpeet) [reply packages/gatsby/src/schema/node-model.js:369]: Ah now you're comment makes sense to me.

File: packages/gatsby/src/schema/node-model.js (hunk 8 above)
Comment (wardpeet) [reply packages/gatsby/src/schema/node-model.js:379]: Wasn't sure if we got 1 item back that's why I asked :)

File: packages/gatsby/src/schema/resolvers.ts (hunk 11 above)
Comment (wardpeet) [reply packages/gatsby/src/schema/resolvers.ts:271]: 👍 Code is hard to follow on who's slicing where but sweet that lmdb does the slicing for us

File: packages/gatsby/src/schema/type-definitions.ts (hunk 12 above)
Comment (wardpeet) [reply packages/gatsby/src/schema/type-definitions.ts:29]: For graphql it's fine but if the connection is used outside of Gatsby it's rather cumbersome to figure out if it's a promise or not but I think we only use these pieces during graphql resolving, right?

Comment (vladar) [reply packages/gatsby/src/schema/type-definitions.ts:29]: Yeah, this is only used for GraphQL resolvers right now. But you are right, I think we can change it to:

```js
totalCount: (() => number | Promise<number>)
```

This will work the same with `await` but graphql will pick the fast sync route when the returned value is number.

File: packages/gatsby/src/schema/resolvers.ts (hunk 9 above)
Comment (vladar) [reply packages/gatsby/src/schema/resolvers.ts:93]: Added a comment about stats usage

File: packages/gatsby/src/schema/resolvers.ts (hunk 11 above)
Comment (vladar) [reply packages/gatsby/src/schema/resolvers.ts]: Yeah, pagination logic with those tricks (like adding additional elements to the left and right) is often puzzling but I've added some tests to demonstrate intended usage. Happy to make it clearer if you have any suggestions 😉

**Review by wardpeet — APPROVED** [review]
